pyk/src/pyk/kast/outer_parser.py

```diff
@@ -62,11 +63,13 @@ class OuterParser:
     _lexer: Iterator[Token]
     _la: Token
     _la2: Token
+    _source: str
 
-    def __init__(self, it: Iterable[str]):
+    def __init__(self, it: Iterable[str], source: str = ''):
         self._lexer = outer_lexer(it)
         self._la = next(self._lexer)
         self._la2 = next(self._lexer, _EOF_TOKEN)
+        self._source = source
```

Comment:
Consider attaching the source in the AST -> KAst transformation instead.

Reply:
I prefer that it's available here in the OuterParser, it can be used to print the source file in the case of a parsing error (see 1ffd6820cb63aceee1c7a8ebbccd6d6fcdde2985)

Reply:
Right, but the parser operates on a stream of characters, source file is not a notion on its abstraction level.

You can still include the source file in error messages with error translation:

```python
try:
    definition = OuterParser(path.read_text()).definition()
except ValueError as err:
    raise ValueError(f'Parse error: {path}: {err}') from err
```

Reply:
On the other hand, it indeed makes sense to include the path in the `Ast` along with the location as you did, so I accept your argument.

In that case however, instead of having an optional parameter for the `Path`, consider a polymorphic `__init__` that either takes an `Iterable[str]` or a `Path`.

Reply:
Do you mean taking a `Path` instead of an `Iterable[str]` for the source text? Then the `OuterParser` would have to handle `.md` files.

I'm adding automerge, so if I didn't understand what you mean we can follow up on it in a later PR.